raster-foundry/raster-foundry
# Improve error-handling with sharing by email

## Overview

This PR makes a few changes to how exceptional cases are handled when sharing annotation projects with users by email:

- It checks for a user's prior existence in Auth0 by email before trying to create a user with that email address. If a user exists, it proceeds with that `Auth0User`.
- it stops error-ing when trying to add a permission to a list of ACRs that is already present. Instead, it returns the existing list, since the db already reflects the state indicated by the user's request
- it makes potential errors in `replacePermissions` explicit in the type so that it'll at least be obvious where errors are coming from if we run into more exceptional cases in the future

### Checklist


## Notes

This doesn't really change what happens in error cases -- but it does fix what occurs in several previous unexpected errors

## Testing Instructions

- assemble api server
- `docker-compose up nginx-api api-server memcached` (doing this explicitly so that `ctrl+c` will also kill memcached -- it saves a step later)
- get a dev user JWT and `export JWT_AUTH_TOKEN=<your token>`
- `echo '{"email": "bogususer@bogus.com"}' | http --auth-type=jwt :9000/api/annotation-projects/6e35b51e-4d67-4cb0-bf23-ca265ef49b24/share` -- this user already exists in Auth0 but will not exist in your local db. This should succeed.
- delete the created user in your db -- you can find them by email. Note that this will not delete the ACRs, since the ACRs column doesn't have any way to keep track of changes in the users table
- repeat the request above -- it should succeed again

Closes #5333
